### PR TITLE
Implement inheritance using in-file modelines

### PR DESCRIPTION
**Pull request by vigoux**
This implements neovim/neovim#13059 (comment)

This behaves like modelines and remove the use of the base_language map.
Also, this allows to fine-tune what we actually want to include per
query, which is better IMO.

One other thing to note to is that using this mechanism, one can just write queries for his language, without knowing anything about lua, which is great.

I would really like everyone's approval on this, to be sure I don't break anything.

### CODE REVIEW

File: lua/nvim-treesitter/query.lua
Comment (stsewd) [inline lua/nvim-treesitter/query.lua:105]: this was changed so the check would be fast, we are parsing all queries just to check its existent.

**Comment by theHamsta** [comment]
> I would really like everyone's approval on this, to be sure I don't break anything.

This could break the JSX/TSX issue which is why the current implementation works the way it is. The reason with that specific language family (probably only with this language family) is that the extension chain does not depend on the query files but on the the language. The JSX nodes must not be in typescript queries but in TSX queries while TSX queries and typescript are the same.

File: lua/nvim-treesitter/query.lua
Comment (vigoux) [reply lua/nvim-treesitter/query.lua:105]: Are you sure about that ? In this version, we are only getting the list of actual query _files_ ?

**Comment by vigoux** [comment]
Tests are passing because query files are not actually included.

File: lua/nvim-treesitter/query.lua
Comment (stsewd) [reply lua/nvim-treesitter/query.lua:105]: oh I see. Well, in the previous version we were only stopping at the first found file (false in the second arg), here we are iterating over all query files found, not sure how much impact that will have with the performance

Comment (theHamsta) [reply lua/nvim-treesitter/query.lua:113]: I would probably be good to cache the result of `get_query` or do we want to react in case of the files changed?

**Comment by vigoux** [comment]
Should be good now. I added a way to optionnal inherit queries (that is, the language will not be included if the query is not the "root" query).

File: lua/nvim-treesitter/query.lua
Comment (kyazdani42) [inline lua/nvim-treesitter/query.lua:92]: what is this variable ? i don't see it defined

File: lua/nvim-treesitter/query.lua
Comment (vigoux) [reply lua/nvim-treesitter/query.lua:113]: We do want to react in case the files changed.

Comment (theHamsta) [reply lua/nvim-treesitter/query.lua:113]: but read and parse the file every time we make a query?  Most users don't write queries and most of the time you don't change the queries.

We could just cache flush the query when the query file type is modified from within vim.

File: lua/nvim-treesitter/query.lua
Comment (theHamsta) [inline lua/nvim-treesitter/query.lua:85]: Allow multiple `;`? Would be nice if @steelsojka could highlight the magic of this comment in the query parser.

File: queries/tsx/locals.scm
Comment (theHamsta) [inline queries/tsx/locals.scm:1]: and javascript

File: queries/tsx/highlights.scm
Comment (theHamsta) [inline queries/tsx/highlights.scm:1]: ~~and javascript. also the order was fine-tuned in some way, I can't recall~~

File: lua/nvim-treesitter/query.lua
Comment (vigoux) [reply lua/nvim-treesitter/query.lua:85]: Yeah that would be nice

File: queries/tsx/highlights.scm
Comment (vigoux) [reply queries/tsx/highlights.scm:1]: Exactly

File: lua/nvim-treesitter/query.lua
Comment (vigoux) [reply lua/nvim-treesitter/query.lua]: Addressed in latest push.

File: lua/nvim-treesitter/query.lua
Comment (vigoux) [reply lua/nvim-treesitter/query.lua:115]: Aren't queries already cached ? If it's not the case, that is worth investigating and implementing in another PR.

Comment (theHamsta) [reply lua/nvim-treesitter/query.lua:115]: queries are cached until the next edit. If you edit, the query files are loaded from file system again. You can experiment with locals/textobjects. Remove them from a query file they will stop working after the next textedit in a file

Comment (vigoux) [reply lua/nvim-treesitter/query.lua:115]: But that's what we want right ?

Comment (theHamsta) [reply lua/nvim-treesitter/query.lua:115]: That's just unnecessary work. It will recurse the whole filesystem, parse the queries, parse the regexes on every keystroke when we use stuff like `is?`. Right now it only does this on `CursorHold` and when you trigger a textobject which seems to be ok. Maybe we could offer a cached variant for someone who wants less overhead for his application.

Comment (vigoux) [reply lua/nvim-treesitter/query.lua:115]: That's effectively a performance issue. Let's open an issue to track the discussion on this, as this is not a blocker of this PR imo (the behaviour is not changed by the PR).

File: lua/nvim-treesitter/query.lua
Comment (stsewd) [inline lua/nvim-treesitter/query.lua:82]: This should be extended to mention the `(lang)` syntax

**Review by stsewd — COMMENTED** [review]
We should document this new thing somewhere

**Comment by theHamsta** [comment]
We should implement this feature for `nvim-treesitter-textobjects` before we merge.

File: lua/nvim-treesitter/query.lua
Comment (theHamsta) [reply lua/nvim-treesitter/query.lua]: you must change or remove the SHA in the lockfile (until we have 0.17.1 nvim we can make a general update of the lockfile)

**Comment by vigoux** [comment]
Why do you need to implement that ? You mean adding the modelines to `nvim-treesitter-textobjects` ?

**Comment by theHamsta** [comment]
> You mean adding the modelines to nvim-treesitter-textobjects ?

or else the plugin will be broken.

**Comment by vigoux** [comment]
I sounds strange to block this PR waiting for another PR in another repo, but as you wish.

**Comment by stsewd** [comment]
I think If the repo is something we maintain (under the same org), we should try to avoid breaking them

File: CONTRIBUTING.md
Comment (theHamsta) [inline CONTRIBUTING.md:71]: *following

**Comment by theHamsta** [comment]
I don't know why want to replace a 5 lines table by a config that's scattered over dozens of files. But if we need to, please give us some time to adapt to the change (copy the redundant config).

One advantage of the modelines is that someone who sees a query for the first time might understand that there are also queries in another file. But there are also disadvantages. We don't have to merge immediately, especially when a user-invisible change can stay user-invisible across all plugins. I will add the headings tomorrow.

**Comment by theHamsta** [comment]
All in all it's probably nicer to understand with modelines. Even if it means some copying for us.

**Comment by vigoux** [comment]
I agree that copying is a bit less nice, but we can then imagine people adding small, generic queries (matching identifiers for example), that are "sourced" in multiple different languages.
Modelines are more flexible IMO, but they effectively cause some duplication (at start).
They'll also alow more flexibility for queries (i.e. having special textobjects for typescript, while the highlights are including javascript ones)

**Comment by theHamsta** [comment]
Seems to work :-). One remark: it's pretty easy to forget the : and to find out what's wrong. Also **no trailing whitespaces**

**Comment by vigoux** [comment]
Lets add those two remarks then.

File: CONTRIBUTING.md
Comment (stsewd) [inline CONTRIBUTING.md:79]:
```suggestion
query you are writing to include it too, you can mark the language as optional (by putting it
```

**Review by theHamsta — APPROVED** [review]
*following (only the typo in contributing.md). Rest LGTM!

I was wondering whether inheritance could also replace the `after` queries, but it probably gets even more complicated when you can inherit from the next rtfile from that language and nobody want to reoder their runtime path for this.

**Comment by vigoux** [comment]
Rebased and adressed the typo, I'll merge once test pass

**Pull request merged**
